Type annotated quoters

winterland1989 opened the pull request:
~~[Rendered](https://github.com/winterland1989/ghc-proposals/blob/type-annotated-quoters/proposals/0000-type-annotated-quoters.rst)~~

[Rendered as rewritten by @isovector](https://github.com/ghc-proposals/ghc-proposals/blob/eeeee74fdfd745a7906d237be112715323a7f887/proposals/0000-type-annotated-quoters.md)

<!-- probot = {"1313345":{"who":"isovector","what":"wants to accept this","when":"2019-07-31T09:00:00.000Z"}} -->

gelisam commented:
How about reusing the existing phantom typed `TExp` instead of introducing a second one? That is, let's use `TExp a` instead of `TExp`:

    data QuasiQuoterExp a = QuasiQuoterExp (String -> Q (TExp a))

winterland1989 commented:
@gelisam  I remember reading somewhere that untyped template is needed because typed one are not always work, but i'm not sure about this. Of course if we can reuse `TExp a`, then yes please. I think i need some answers from GHC HQ.

gelisam commented:
>  remember reading somewhere that untyped template is needed because typed one are not always work, but i'm not sure about this.

Untyped TH is needed because it is more powerful than typed TH. For example, in [category-syntax](https://github.com/gelisam/category-syntax) my macro takes in code which does not type-check, so it cannot have type `TExp a` for any `a`. It can have type `Exp` though.

[`TExp`](https://hackage.haskell.org/package/template-haskell-2.12.0.0/docs/Language-Haskell-TH-Syntax.html#t:TExp) is just a wrapper around `Exp` with a phantom type:

    newtype TExp a = TExp Exp

So using `String -> Q (TExp a)` instead of `String -> Q Exp` would certainly not limit the expressivity of type-annotated quasiquoters.

goldfirere commented:
I'm not much of a consumer of quasi-quoters, so I don't have an informed opinion as to whether this is a good idea.

I do agree that it should use `TExp`.

simonpj commented:
This proposal makes sense to me.  It's just a natural extension of the entire Typed Template Haskell idea, when `TExp` was first introduced.  With that in mind can we call it `TQuasiQuoter a` please?  Just like `Exp` vs `TExp`.

It's much more than just documentation!!  Consider
```
qq :: TQuasiQuoter Char

blah = [|qq| unicode 78 |] && True
```
With the existing quasi-quote machinery we'd first have to run `qq`, splice in the resulting syntax tree, and then complain if it didn't typecheck.  With a typed quasi-quoter we can complain right away: `qq` returns a `TExp Char` and the doesn't fit somewhere a `Bool` is needed.   

Worth explaining this in the proposal.

winterland1989 commented:
@simonpj 

> With the existing quasi-quote machinery we'd first have to run qq, splice in the resulting syntax tree, and then complain if it didn't typecheck. With a typed quasi-quoter we can complain right away: qq returns a TExp Char and the doesn't fit somewhere a Bool is needed.

This's a good point, i have added to `Effect and Interactions` section of the proposal, also updated to use `TQuasiQuoter a` instead of `QuasiQuoterExp a`. I personally don't have any problems with either name, hope not getting too much bikeshedding here.

winterland1989 commented:
@gbaz , `TQuasiQuoter a` should be run during the renamer just as current QQ, the only difference is that we're splicing `TExp a` AST node instead of `Exp` node. The benefit that simon described shoud be a nature result of using `TExp a` node. i.e. the major part of this proposal has been in place, we only have to add an user interface to support it.

simonpj commented:
>  TQuasiQuoter a should be run during the renamer just as current QQ, 

I don't think so.  Just like `TExp` it should run in the type checker.   Somewhere we have to check that the type matches the place it is spliced in, and the type checker is the only place to do that. I'm arguing for a tight connection with `TExp`.

tejon commented:
Pardon the tangent, especially since I'm only following half of this; but I can't help wondering if there might be some potential cross-pollination with [Hackett](https://github.com/lexi-lambda/hackett) here, since its whole purpose for existence is to unify type checking with macro expansion. That's not the same as unifying with TH, so it's not directly applicable to this proposal; but I think it solves the same problem?

Of course, the impedance mismatch might simply be too strong. But @lexi-lambda has mentioned that if it were more viable/convenient, she would like GHC as a Hackett backend, at which point Hackett modules should be importable from Haskell, right? What's in the way of that? Is it harder than working out all the kinks and details here?

Again, pardon the derailment, this isn't really the right place to bring this up. But it is the right context and audience, and I'm not sure where else to find that!

winterland1989 commented:
@simonpj I'm definitely wrong on how a `TExp a` node got spliced, thanks for correcting me on this! I have added some notes under the proposal's `Proposed Change Specification` section to make this point clear.

winterland1989 commented:
@nomeata I think this is ready for committee review ; )

nomeata commented:
Manuel volunteered to shepherd this.

winterland1989 commented:
OK, let's see if my friends @sighingnow got some time.

mchakravarty commented:
@winterland1989 I know that you asked to submit this proposal to the committee already, but, if you don't mind, I'd like to clarify a few points first to make sure the proposal is really ready.

Firstly, in TH, we use `[|...|]` for untyped quotations and `[||...||]` for typed quotations. So, in your example, don't we have to write this?

```
blah = [qq|| unicode 78 ||] && True
```

Secondly, I understand that this goes beyond your proposal, but the obvious next question here is whether we can have some type information flow between the quasi-quoter and GHC's type checker. After all, we are running these quasi-quoters now as a part of the type checker.

It would be a pity to accept this proposal now as is and only to discover later that it could have been generalised to cover this information flow and then having to worry about backwards compatibility.

gbaz commented:
Manuel: What sort of information flow in what direction would you be imagining? I've often thought we could do more in the case of standard typed TH splices (c.f. https://github.com/ghc-proposals/ghc-proposals/pull/124#issuecomment-382253458 for example), but in the case of a quasi-quoter what information would you like to be able toou have access to that this proposal currently doesn't allow?

I know your inline-c stuff would be a good use-case for this sort of stuff, so I'm sure you have some further ideas here that would be worth sketching.

winterland1989 commented:
@mchakravarty This proposal is more about extending quasi-quotation, i.e. use `foo :: TQuasiQuote a` with `[foo|...|]` syntax, so it should not have any effect on previous `[e|| .. ||]` typed quotes.  But i do see your point: should we reuse the double pipe syntax? I'd rather say no, since it's not real necessary. Let's keep proposed syntax as it is.

> Secondly, I understand that this goes beyond your proposal, but the obvious next question here is whether we can have some type information flow between the quasi-quoter and GHC's type checker. After all, we are running these quasi-quoters now as a part of the type checker.

Let's the implementer find out then ; )

mchakravarty commented:
@gbaz wrote,
> Manuel: What sort of information flow in what direction would you be imagining? I've often thought we could do more in the case of standard typed TH splices (c.f. #124 (comment) for example), but in the case of a quasi-quoter what information would you like to be able toou have access to that this proposal currently doesn't allow?

What we would really want is for the type information gathered by the quasi-quoted language's type checker to be propagated to the meta-program and vice versa. This is what happens in Typed TH, after all.

In case of 
```
blah = [qq|| unicode 78 ||] && True
```
it is trivial, because `qq` already fixes the type to be a `Char`. For an even slightly richer qq'ed language this is no longer the case.

I appreciate that this is outside the scope of this proposal, but I'd prefer this proposal to be a first step in the right direction (as opposed to making it harder to be more general later on).

Does that make sense?

mchakravarty commented:
@winterland1989 wrote,
> @mchakravarty This proposal is more about extending quasi-quotation, i.e. use `foo :: TQuasiQuote a` with `[foo|...|]` syntax, so it should not have any effect on previous `[e|| .. ||]` typed quotes. But i do see your point: should we reuse the double pipe syntax? I'd rather say no, since it's not real necessary. Let's keep proposed syntax as it is.

I understand that it is not necessary, but the question is whether it would be more consistent language design.

> Secondly, I understand that this goes beyond your proposal, but the obvious next question here is whether we can have some type information flow between the quasi-quoter and GHC's type checker. After all, we are running these quasi-quoters now as a part of the type checker.
> Let's the implementer find out then ; )

Sorry, but this is not how the proposal process is supposed to work. We want to explore the implications of a proposal before deciding whether to accept it.

I appreciate that you are aiming for a simple extension and may not want to go the full way to propagating type information. That is fine and a reasonable first step. However, it does require us to think hard about the exact approach so that the next step can be a **seamless extension** from what we do now.

gbaz commented:
> This is what happens in Typed TH, after all.

I was trying to understand what you were getting at, and I think I have an idea now. I suppose the one thing you may be thinking of is that a typed _quote_ is attributed type `TExp a` where the `a` reflects the type of the quoted expression. So perhaps what you are suggesting is that similarly we might imagine typed quasi-quoters that aren't fixed, but are instead generative with regards to the `a` in the `TExp` they generate. I agree this is a very interesting idea! 

So what if a typed quasiquoter actually did _not_ yield a `TExp a` but instead a pair of an `Exp` and a `Type`? And the compiler did the "smart" thing and converted that into an appropriate `TExp` of the appropriate type. Everything is running in the type-checker phase, so it certainly _could_.

The problem is now we've lost the original motivation for the proposal :-/

> The type QuasiQuoter is rather uninformative, especially when it's an expression quoter. Those QuasiQuoters appeared in documents on hackage not only didn't give users useful information, but also scared beginners away.

Instead, we get more of the benefit suggested by SPJ -- earlier detection of bad type assignments. But this earlier detection I don't think is worth that much. The code we generate still is only as correct as we "know" our joint calculation of expression and type to properly match is, so could still fail in the checking phase, for example.

Now, if we had something like what I suggested in the thing linked above, we might get further. In particular, if we were in a situation where we were expected to produce a `TExp a` for some `a`, would it be possible to gather and make available all the typing constraints in the ambient environment with regards to that `a` (via the `Q` monad)? This would be a strict improvement not only to typed quasiquoters but typed quoters in general -- and furthermore it would be absolutely compatible with the proposal as described here, since it would involve extending the `Q` monad when invoked in the typechecking phase, not any syntactic changes or changes to datatypes...

mchakravarty commented:
@winterland1989 What are we going to do here now? Is your standpoint that you want to submit as is without further discussing the two points I last raised?

winterland1989 commented:
@mchakravarty , I'm considering a more drastic change: what if we generalize `[foo| ... |]`, so that it's more tolerant on types of quoters  it can accept?  e.g. a `TExp a`  can be directly spliced by `[e|bar|]`, like a typed quoters i proposed here.  

Doing so need compiler's support, i'm not sure if this is possible. And it's effectively make `[e|| ... ||]` syntax redundant.

I'm not very clear on your second question though, can you be more specific on which part or step of the compiling?

isovector commented:
A sincere apology from the whole committee about how slow we've been on this proposal! Let's fix that!

I'm a :+1: on this proposal; it solves a concrete problem and adds type-safety to a place that we currently don't have any. A few notes:

* I think the `[foo|| blah ||]` syntax is preferable to `[foo| blah |]`. It's confluent with the `TExp` quotes, serves as a lamp post to casual readers that this splice is typesafe, and (I suspect) would simplify the implementation.
* If we added a few more motivating examples of things that are hard to do safely today, I think it would be easy to convince the committee. We could steal a few from the `compile-time literal values` proposal. A few off the type of my head: ensuring that a `WordX` won't under/overflow, and that a `NonEmpty` is in fact non-empty. I'm particularly interested in the `NonEmpty` example, as I want to have a clear idea how these things will interact with polymorphism.

goldfirere commented:
What about @mchakravarty's concerns about information flow? Perhaps the parser of the quasiquoted string wants to know whether it's trying to produce an `Int` or a `Double`. Does this design allow for that? I don't think so. Actually, it seems that you wouldn't be able to write one parser that could work for more than one type, at all, here.

On the other hand, if we had

```
data TQuasiQuoter a = MkTQQ (Typeable a => String -> Q (TExp a))
```

we'd be closer to the mark. However, that won't work in cases where we want to *infer* the type of the resulting expression. So maybe we need

```
data TQuasiQuoter a = MkTQQ (Maybe (TypeRep a) -> String -> Q (TExp a))
```

That's a bit closer, but now the `Nothing`/`Just` choice will depend on the vagaries of type inference. So we'd have to be careful to supply `Just` only when propagating a type down via bidirectional type inference. I think that would work, but I want someone else to check my reasoning here and to see if there's a simpler design available.

isovector commented:
> What about @mchakravarty's concerns about information flow? Perhaps the parser of the quasiquoted string wants to know whether it's trying to produce an Int or a Double. Does this design allow for that? I don't think so. Actually, it seems that you wouldn't be able to write one parser that could work for more than one type, at all, here.

All good questions. I think the easiest way to hash this out is to have some concrete examples of things we'd like, and to go from there.

Ericson2314 commented:
Stuck macros http://gelisam.com/files/stuck-macros.pdf / https://www.youtube.com/watch?v=nUvKoG_V_U0&list=PLNoHgLVTxtaorTczyo8NA3tg_vK8WC5rD&index=9&t=0s from Compose 2019 is the right idea for this stuff. CC @gelisam

mpickering commented:
1. I think that you need a different syntax for typed qqs. The untyped and typed variants now result in different types so using the same syntax seems problematic.

2. This question about the flow of information is really a much tricker question about how types should interact with top-level splices. The behavior here is not currently very well specified/formalised (something I plan to clearly state in the near future). I think we should stay out of this territory for now
and deal with the general issue first.

For example, quantified type variables are in scope across a top-level splice.

```
qid :: forall a . Q (TExp (a -> a))
qid = [|| id ||]

res :: forall a . a -> a
res = $(qid @a)
``` 

but not variables, constraints or implicit parameters (the relevant arguments). I think types only work by accident currently.

Attempting to dispatch based on a dictionary is rejected:

```
{-# LANGUAGE TemplateHaskell #-}                                                    
{-# LANGUAGE TypeApplications #-}                                                   
{-# LANGUAGE ScopedTypeVariables #-}                                                
{-# LANGUAGE GADTs #-}                                                              
module E where                                                                      
                                                                                    
import Type.Reflection                                                              
                                                                                    
foo :: forall a . Typeable a => a                                                   
foo = $$( case typeRep @a `eqTypeRep` typeRep @() of                                
            Just HRefl -> [|| () ||]                                                
            _ -> error "type" )
```

```
E.hs:10:16: error:                                       
    • No instance for (Typeable a) arising from a use of ‘typeRep’
    • In the first argument of ‘eqTypeRep’, namely ‘typeRep @a’
      In the expression: typeRep @a `eqTypeRep` typeRep @()
      In the expression:
        case typeRep @a `eqTypeRep` typeRep @() of
          Just HRefl -> [|| () ||]
          _ -> error "type"
   |
10 | foo = $$( case typeRep @a `eqTypeRep` typeRep @() of
   |                ^^^^^^^^^^
Failed, no modules loaded.    
Prelude> Leaving GHCi.  
```

I'm of course still free to use type class methods with concrete types.

```
foo :: ()                                                                           
foo = $$( case typeRep @() `eqTypeRep` typeRep @() of                               
            Just HRefl -> [|| () ||]                                                
            _ -> error "type" ) 
```

and, crucially, I can define a polymorphic function which can be used inside a top-level quote as long as the constraints are solved. 

```
module F where

poly :: forall a . Typeable a => Q (TExp a)                                         
poly =                                                                              
  case typeRep @a `eqTypeRep` typeRep @() of                                        
            Just HRefl -> [|| () ||]                                                
            _ -> error "type"   

module E where

foo :: ()
foo = $$( poly )
```
Seeing as QQs desugar to a top-level splice we should be consistent with their treatment. This means that you can dispatch based on a type class if the constraint is solved but constraints are not abstracted over. I think this behaviour satisfies both @mchakravarty and @goldfirere in a quoter which
can use information about its context. 

So I think the proposal should be accepted with the additional of new syntax. 

(BTW, I am also happy to implement this if the proposal gets accepted)

gelisam commented:
> "Stuck Macros" [...] from Compose 2019 is the right idea for this stuff.  
> [...]  
> This question about the flow of information is really a much tricker question about how types should interact with top-level splices.

Would you like me to write a proposal for adding stuck macros to GHC? I expect it would be a pretty big change, so I will need help from people who are familiar with the macro-expansion and type-checking phases of GHC in order to better understand the ramifications of such a change. But at least I could get the ball rolling by filling in the motivation and proposed changes section.

gelisam commented:
> Attempting to dispatch based on a dictionary is rejected:
> 
>     foo :: forall a . Typeable a => a                                                   
>     foo = $$( case typeRep @a `eqTypeRep` typeRep @() of                                
>                 Just HRefl -> [|| () ||]                                                
>                 _ -> error "type" )

That makes sense: the code is being generated at compile-time, but the dictionary is only received at runtime, so of course we cannot obtain the `TypeRep` at compile-time. We can, of course, generate a `TExp (forall a. Typeable a => a)` (with the appropriate `newtype` to avoid impredicative polymorphism) which examines the `TypeRep` at runtime.

    {-# LANGUAGE GADTs, RankNTypes, ScopedTypeVariables, TemplateHaskell, TypeApplications #-}
    import Language.Haskell.TH
    import Type.Reflection
    
    newtype AnyTypeable = AnyTypeable
      { runAnyTypeable :: forall a. Typeable a => a }
    
    -- |
    -- >>> $$(foo) :: ()
    -- ()
    foo :: Q (TExp AnyTypeable)
    foo = [|| let x :: forall a. Typeable a => a
                  x = case typeRep @a `eqTypeRep` typeRep @() of
                        Just HRefl -> ()
                        _ -> error "type"
              in AnyTypeable x ||]

mpickering commented:
@gelisam Is there a paper or document which explains what the idea is? I've never heard of a "stuck macro" before.

gelisam commented:
> @gelisam Is there a paper or document which explains what the idea is? I've never heard of a "stuck macro" before.

It's a term I came up with. I didn't write a paper, the only information which is available online is the slides and the talk which has already been linked above, plus [the github issue in which I propose adding stuck macros to Hackett](https://github.com/lexi-lambda/hackett/issues/61). However, at Compose, @david-christiansen seemed to know more about my talk's topic than I did myself, so maybe I simply rediscovered something which already exists under a different name? If so, David, do you know about the relevant literature?

mpickering commented:
@gelisam I just glanced through your slides and it seems you only talked about untyped metaprogramming, is there a typed variant as well? I will listen to your talk as well.

david-christiansen commented:
I don't know of existing literature on this topic. I've been pondering integrating macro expansion and type checking for some time now, and I have a prototype implementation of a language that does it, but its type checker doesn't do unification. To add unification, I had arrived at a similar idea to yours, but I hadn't yet implemented it to see if it's actually practical.

The closest existing thing to this that I know of is in [Agda's metaprogramming system](https://agda.readthedocs.io/en/v2.5.2/language/reflection.html), which has this feature:
```agda
  -- Block a type checking computation on a metavariable. This will abort
  -- the computation and restart it (from the beginning) when the
  -- metavariable is solved.
  blockOnMeta : ∀ {a} {A : Set a} → Meta → TC A
```
Your proposal is similar, except that you've proposed a dynamic semantics that captures a continuation to restart the computation where it left off instead of restarting, and you've proposed an API where blocking is an implicit result of unification rather than an explicit command given by the user. Right, @gelisam?

Ericson2314 commented:
@mpickering FWIW I consider the stuck macros stuff potentially a *replacement* for typed meta-programming as we have it today. I say if we start with completely stage-separated https://github.com/ghc-proposals/ghc-proposals/pull/243, then stuck macros allow one to very precisely opt into blocking on more information being gotten from the next stage.

mpickering commented:
Thanks @gelisam I have watched your talk now. It seems that the most natural way to implement a "stuck macro" is to closely integrate it into the constraint solver as you want to "block" until a unification variable is solved. 

However, most of your examples from the first part of the talk (for example `pullMaybe`) could be implemented already in GHC using type families. It does seem though that there is potentially more power here though, as the type checking and expansion processes are interleaved, but it's not clear exactly what other uses you have in mind.

Here's my implementation of the main example from your talk. 

https://gist.github.com/mpickering/c681d1c89b28bf4a8d39f06cead12835

I could have implemented something similar combining together type classes and quotations but decided to use type classes to do the code generation this time. 

Anyway, I don't think this much to do with the original proposal so we shouldn't derail it any more and work on this idea later if we're still inclined.

isovector commented:
@Ericson2314 @mpickering @gelisam @david-christiansen thanks for your input! If you think there is merit in exploring these ideas further, we'd welcome a separate proposal!

@winterland1989 To reiterate, if you're still interested in pursuing this proposal, I suggest we work together on coming up with several motivating examples. In particular I'm interested in how this feature would interact with polymorphism, constraints, and impredicativity.

tejon commented:
> integrating macro expansion and type checking

I'm significantly out of my depth, but this sounds like what @lexi-lambda is working on with [Hackett:](https://github.com/lexi-lambda/hackett) (and that *is* based on [a paper](http://www.ccs.neu.edu/home/stchang/pubs/ckg-popl2017.pdf)).

mpickering commented:
@isovector Shouldn't we just accept this proposal and move on? It was opened over a year ago.

Ericson2314 commented:
Yeah to flip it around, I think *since* the limitations the stuck macro work attacks apply to typed TH generally, this proposal is marginally innocent; there is no problem with "polymorphism, constraints, and impredicativity" unique to this and not applying to typed TH as a whole.

Ericson2314 commented:
Oh, is there any reason why pattern quasi-quoters cannot also be typed? (I suppose there is or we would have typed pattern splices already?)

david-christiansen commented:
@tejon The work is indeed closely related - this isn't the place to get into the relationship, though. Email me if you want to hear more.

@isovector I think that the "stuck macro" idea is not ready for something like a GHC proposal - hopefully it will be someday, however.

And I agree with the consensus here that, while @gelisam's idea is interesting and worth exploring elsewhere, it's orthogonal to this proposal.

isovector commented:
@mpickering yeah, I'm inclined to agree. I reread https://github.com/ghc-proposals/ghc-proposals/pull/125#issuecomment-510424321 and https://github.com/ghc-proposals/ghc-proposals/pull/125#issuecomment-510480380 and find them sufficiently convincing that we can deal with polymorphism and impredicativity. 

I'm going to recommend this proposal for acceptance with the change that the syntax be `[qq|| foo ||]`. My rationale is thus:

* It's a small, natural extension to quasiquoters and typed expression quotes. 
* It gives us a nice solution for #124, which is a very real problem in industrial-world Haskell
* Statically known dictionaries can be used without any issues, even for polymorphic variables (see `checkedNum` below)
* Typed QQs can return GADTs/newtypes packing foralls and Dicts to get around impredacitivity concerns.
* This proposal has been "pending committee review" for more than a year and deserves a prompt decision, whether yea or nay.

```haskell
checkedNum
    :: forall a
     . (Bounded a, Num a, Integral a, Typeable a)
    => String
    -> Q (TExp a)
checkedNum str =
  if minVal <= val && val <= maxVal
      then [|| fromInteger val ||]
      else fail $ mconcat
        [ show val
        , " is out of bounds for "
        , show (typeRep $ Proxy @a)
        ]
  where
    minVal = fromIntegral $ minBound @a
    maxVal = fromIntegral $ maxBound @a
    val = read @Integer str

ok :: Word8
ok = $$(checkedNum "255")

bad :: Word8
bad = $$(checkedNum "256")
--     • 256 is out of bounds for Word8
--     • In the Template Haskell splice $$(checkedNum "256")
--       In the expression: $$(checkedNum "256") :: Word8
--       In an equation for ‘ok’: ok = $$(checkedNum "256") :: Word8
```

isovector commented:
**This proposal will now be discussed by the committee. We welcome all authors to join the discussion, but kindly ask others to refrain from posting**

goldfirere commented:
There's been quite a lot of conversation that has not yet worked its way into the proposal proper. I've been keeping up, but perhaps other members of this committee have not. If we (understandably) don't wish to burden @winterland1989 with requiring an update, could you @isovector summarize the state of affairs? I do think someone will have to update the proposal before its eventual acceptance.

In particular, @mpickering's https://github.com/ghc-proposals/ghc-proposals/pull/125#issuecomment-510424321 is very convincing that we can capture the "information flow" with the current design, with one tweak: the type of a quasiquoter needn't be `TQuasiQuoter a`, but may be a subtype of `TQuasiQuoter a`. That is, we need to allow `qq :: Typeable a => TQuasiQuoter a` to adapt @mpickering's example. (Other constraints would be welcome there, too -- it's just that `Typeable` is very useful in that spot.)

goldfirere commented:
Oh, and I'm in support, with the `[qq|| ... ||]` syntax.

isovector commented:
@winterland1989 are you comfortable with me editing the proposal text on your behalf? I don't want to step on your toes, nor do I want to force you to get all of this back into working memory after a year. If I don't hear from you in two days, I will assume you're OK with me making edits.

@goldfirere If I understand you correctly, the distinction you're drawing in https://github.com/ghc-proposals/ghc-proposals/pull/125#issuecomment-510629909 is this: a typed qq returning `TExp a` _may_ have a constraint on `a`, and _if it does_, such a constraint must be dischargeable at compile time at the call site. Is that correct?

simonpj commented:
> There's been quite a lot of conversation that has not yet worked its way into the proposal proper

I for one have not been following the discussion over the last few days in detail.  But in any case, I think it is good practice for the committee to consider the proposal, on it own merites in its final stand-alone form, rather tthan the proposal as modified by a stream of discussion.  I'm not saying that's true of this one.   Just saying that I much prefer to review an actual proposal that has been refined in the light of feedback and that the author now wants the committee to consider.  Thank isovector for offering!

simonpj commented:
At first I thought this was a relatively straightforward proposal, but reading this thread I'm suddenly much less sure!

Suppose we try to define
```
qq1 :: TQuasiQuoter (Char -> Char)
qq1 = TQuasiQuoter  pCharFun
```
where `pCharFun :: String -> Q (TExp (Char -> Char))`.    Now let's write a possible `pCharFun`.  Um.  How do I build a `TExpr (Char -> Char)`.  The only way I know is with a typed quote like `[|| \x -> ord x||]`.   But that doesn't take a string as input.

Can someoen give a compelling example of a typed parser like `pCharFun`?   You don't need the proposed extension to write it -- you should be able to write it today.

------------------------

Second question.  Is it possible to write a polymorphic typed quasiquoter?  Or must it be entirely monomorphic?   A polymorphic one might be something like
```
qq2 :: forall a. TQuasiQuoter (a->a)
```
Can we write one down as a demo?  I'm not sure how to.
---------------

Third question: do we have any "live customers", that is, people saying "if we had this I'd use it".  If so, we can check that the proposed extension is capable of dealing with their applications.

isovector commented:
@simonpj here are a few "qq"s I whipped up in 8.6.3:

Question 1: We don't need to bind the string inside the typed exp quote:

```haskell
-- Gives back a function that will permute a character with its subsequent character in a string.
qq1 :: String -> Q (TExp (Char -> Char))
qq1 str = [|| \c ->
    case lookup c rotated of
      Just c' -> c'
      Nothing -> c ||]
  where
    rotated = zip str (tail str ++ str)
```

It's not the world's most interesting example, but it does indeed work:

```
> fmap $$(qq1 "abcdefgh") "hello world"
"afllo worle"
```

---

Question 2: Your desired type isn't very useful, but we can write `qq2` also:

```haskell
qq2 :: String -> Q (TExp (a -> a))
qq2 = const [|| id ||]
```

For a more interesting example of polymorphism, I provided `checkedNum` above:

```haskell
checkedNum
    :: forall a
     . (Bounded a, Num a, Integral a, Typeable a)
    => String
    -> Q (TExp a)
checkedNum str =
  if minVal <= val && val <= maxVal
      then [|| fromInteger val ||]
      else fail $ mconcat
        [ show val
        , " is out of bounds for "
        , show (typeRep $ Proxy @a)
        ]
  where
    minVal = fromIntegral $ minBound @a
    maxVal = fromIntegral $ maxBound @a
    val = read @Integer str
```

which will ensure literals won't under/overflow for any `Bounded, Num, Integral` type. Furthermore, this is an interesting case that _can't_ be solved today with just `QuasiQuoter`s, since it branches based on the type of `a` it's invoked at. `checkedNum` will warn you about overflows for custom times, which `-Woverflowed-literals` will not.

simonpj commented:
Thanks Sandy,that's helpful.   But all those examples have the flavour of 
* Use a conditional to pick quote 1 or quote 2
* Use lifting to get a parsed constant into quote

This is very far from parsing a rich domain-specific language, which is what quasiquotes are intended for.

In my current understanding, the gain from this new extension is therefore very slight, making typed quasiquotes vastly less useful that untyped ones.  So the pain-to-gain ratio for the proposal seems too high.   My instinct is to let this simmer longer, awaiting a better way.   (Maybe along the lines of Hackett.)

But I could well be missing something.   Perhaps a bunch of people will answer my question 3, saying "yes, this is just want I need".

isovector commented:
I'm tagging a number of hackage authors who have published libraries defining quasiquoters. If I've tagged you here, it means the committee could use your help! In particular we're discussing *typed quasiquoters,* which is to say a value of `TQuasiQuoter a` which when run will produce an `a`. Such a thing would be typechecked at both its definition and at its use.

*Would such a thing be useful to you?* Has the lack of typed quasiquoters gotten in your way? Did it lead to bugs in your implementation?

@sol @mboes @audreyt @finnsson @biegunka @takano-akio @konn @metrix-ai

---

I'd also like to explicitly invite @mpickering who AFAIK has done more thinking about typed splices than anyone I've ever met. Also to @parsonsmatt, whose proposal #124 was the original idea that this proposal spun out of.

---

@simonpj I don't think I understand what you're asking for. What's a sketch of a rich domain-specific language that this proposal would preclude?

mpickering commented:
It is common to write a parser `String -> Maybe a`, this is a "typed parser" in the language of this proposal. If you can do that then you can also write a typed quasiquoter `String -> TExpQ a`. 

For a quasiquoter, how you construct the `TExpQ a` determines how much evaluation happens at compile time. If you construct your typed quasiquoter using `lift . read` then all evaluation happens at compile time. If you construct it using `\s -> [| read $(lift s) |]` then you defer the evaluation to runtime. There is also of course a spectrum in-between these two possibilities. 

For a language with an advanced type system, it seems sensible if the metaprogramming is also typed.

simonpj commented:
> What's a sketch of a rich domain-specific language that this proposal would preclude?

The original paper [Why it's nice to be quoted](https://www.cs.tufts.edu/comp/150FP/archive/geoff-mainland/quasiquoting.pdf) has an excellent exposition.  As a very large example the [PADS project](http://www.padsproj.org/) has a huge domain specific language embedded in Haskell via quasiquotation.

But to take a simple example suppose I wanted to quasiquote a lambda calculus like language with very verbose syntax.  With quasiquotes I can say
```
[lam|  FUNCTION(x) ARROW (PLUS x (STRING "one")) |]
```
Silly I know, but that might be the DSL I want to parse, and turn into Haskell.   I know exactly how I'd do this with untype quasiquotation (see the paper), but I can't see how to begin doing it with this proposal, for the reasons I describe above -- namely that to make a term of type `Q (TExp ty)`, I can only do so with a typed quote or a lifted constant.

isovector commented:
I don't think we disagree here!

Am I correct in saying the issues with your your lambda calculus example are:

1. that the type of the splice depends on which string you're actually parsing
2. you want to emit a binder for `\x ->` and then go on  parsing the `(PLUS ...` stuff, which ensuring you can refer back to `x` in the splice?

1 is a problem for every typed system in the world which doesn't have typechecking as a first class concept, and I'm sure we could get around 2 with an `unsafeTExp :: Exp -> TExp a` function.

I'm quickly getting out of my depth here, but would like remind everyone that this proposal isn't to take away untyped QQs. Just to add typed ones! Typed qqs like my `checkedNum` example, can do things that AFAIK untyped ones can't --- they can branch on the type of the resulting splice!

In my eyes, despite their similarity, the typed and untyped quasiquoters are orthogonal features.

However, I'm inclined that we make a decision here. My recommendation i to accept. If you're still hesitant about the idea, I propose giving a few examples of small, useful DSLs you'd consider this feature to be *incomplete* if it couldn't handle. We can then wait a week for a response, and accept if there's a solution, and reject otherwise. Does that sound reasonable to you?

parsonsmatt commented:
I'd love to see this land. It would allow for safer, more flexible, and more expressive literals than the status quo wrt #124 . In particular, we can provide significantly better error messages when we know more about the type that we're trying to handle.

Imagine a QuasiQuoter that parses a database model definition file, and as it parses the file, it collects information about which database features the backend needs to support. It returns a polymorphic data structure with the form `(SqlSchema a, HasCheckConstraints a, HasForeignKeys a, HasJsonColumns a, ...) => a`. On the parsing side, we have a set of functions `PostgresSchema -> Q [Dec]`, `MysqlSchema -> Q [Dec]`, `MongodbSchema -> Q [Dec]`, `Neo4jSchema -> Q [Dec]`, etc. These functions pick the type of the schema and render out the appropriate definitions for interacting with it.

With an untyped quasiquoter, we get an error message like:

> ```
> Unable to deduce HasCheckConstraints MysqlSchema
> based on a usage of whateaverasdfasdfasdfasdf in some generated code
> ```

Which isn't a great UX. We can improve this now by using a `TypeError` instance for each of the missing functionality classes for each of the database backend types, but that's also bad (y'all thought people complained about the `mtl` instance problem).

If the QuasiQuoter could reflect on the type, then we could know the capabilities missing and report better error messages. Instead of propagating the type class instances to the using side, the type would be available in the quasiquoter itself, and we could provide an error message within the context of the QQ, including which field it was on, which line in the QQ we're on, etc.

isovector commented:
@simonpj gentle ping on this!

simonpj commented:
> that the type of the splice depends on which string you're actually parsing

Yes, that's certianly part of it.   But even if I know the type of the thing I'm pasing, how do I write `parseFun` which is meant to parse this little epxression language?
```
parseFun :: String -> Q (TExpr (Int->Int))
parseFun = ???
```
The only exxample we have seen so far, namely your `qq1` and `checkedNum` involve liftable constants only: booleans or numbers, which you can parse out of the string and then lift into a quote.  But that's an extremly limited class of parsers.

As things stand, I just can't see how this feature will be useful in practice.   Perhaps I'm blind and there are lots of compelling use-cases  --- in which case it should be easy to exhibit some.

yav commented:
Hello,

I think the proposal seems reasonable, and as far as I understand it is a very small delta to the status quo.  I also think that it could be potentially useful for writing DSLs, and I just had some fun writing a small example along the lines of what @simonpj asked for.  Here it is:

```Haskell
module DSL where

import Language.Haskell.TH

data Expr = Fun String Expr | Add Expr Expr | Var String
            deriving Show

type Code a = Q (TExp a)

-- Language quoter
lam :: String -> Code Expr
lam input = case pExpr (words input) of
              Just(e,[]) -> e
              _          -> fail "Parse error"

pExpr :: [String] -> Maybe (Code Expr, [String])
pExpr s = case s of
            "ADD" : s1 ->
               do (a,s2) <- pExpr s1
                  (b,s3) <- pExpr s2
                  pure ([|| Add $$a $$b ||], s3)

            "FUN" : v : "->" : s1 ->
               do (a,s2) <- pExpr s1
                  pure ([|| Fun v $$a ||] , s2)

            x : s1 ->
              pure ([|| Var x ||], s1)

data Val = VFun (Val -> Val) | VInt Int | VErr

compile :: [(String,Code Val)] -> Expr -> Code Val
compile env expr =
  case expr of
    Fun x e -> [|| VFun (\i -> $$(compile ((x, [|| i ||]) : env) e)) ||]

    Var x -> case lookup x env of
               Just i -> i
               _      -> [|| VErr ||]

    Add x y -> [|| case ($$(compile env x), $$(compile env y)) of
                    (VInt x, VInt y) -> VInt (x + y)
                    _ -> VErr ||]

-- example = $$(compile [] $$(lam "FUN X -> ADD X X"))
```
The first part shows how you could use a quoter to get a nicer notation for an AST.   The second part shows how you could compile your DSL into a Haskell expression (this may be part of the quoter, or not, depending on your goals).

As this shows, we can already do all of this stuff with existing typed template Haskell.   As far as I understand, the main benefit of this proposal is that it would let you write:
```Haskell
[lam || "FUN X -> ADD X" ||]
```
instead of
```Haskell
lam "FUN X -> ADD X"
```
The difference is minimal, but the first one has the benefit of supporting multiple lines, and not having to escape quotes, etc.

PS:  should a typed quoter use two bars (like typed TH) or one, as is in the proposal at the moment?

isovector commented:
@simonpj what do you think of @yav's example? I think it, combined with @parsonsmatt and my examples is sufficient to show that this is a useful feature. 

If there are remaining concerns, I'd like to remind everyone that `TExp` is just a newtype over `Exp` with a phantom parameter. Nothing is stopping us from exposing an `unsafeTExp :: Exp -> TExp a`, which would allow the user to unleash the full fury of untyped TH to construct new typed primitives.

Unless anyone strongly disagrees with this, I am going to recommend we accept this proposal one week from today, on July 29.

isovector commented:
@simonpj what do you think of @yav's example? I think it, combined with @parsonsmatt and my examples is sufficient to show that this is a useful feature. 

If there are remaining concerns, I'd like to remind everyone that `TExp` is just a newtype over `Exp` with a phantom parameter. Nothing is stopping us from exposing an `unsafeTExp :: Exp -> TExp a`, which would allow the user to unleash the full fury of untyped TH to construct new typed primitives.

Unless anyone strongly disagrees with this, I am going to recommend we accept this proposal one week from today, on July 29.

nomeata commented:
/remind me that this can be accepted on July 29

goldfirere commented:
I got an email from @simonpj that he was on holiday today. I don't know whether this extends for the whole week. I would love to hear more of his input, though it seems that his primary concern has been addressed.

nomeata commented:
/remind me that @isovector wants to accept this on July 29

nomeata commented:
/remind @isovector to accept this on July 29

simonpj commented:
Iavor's example is indeed helpful.  I now see this proposal primarly as a way to write literal values in user-defined syntax.   All the examples have this form:

* `[qq|| unicode 78 ||]` is user-defined syntax for a `Char` literal.   (This from the proposal.)
* `[qq1|| abcdefgh ||]` is user defined syntax for a `Char->Char` function that does funny rotation.
* `[checkedNum|| 241 ||] is user-defined syntax for an `Integer` whose bouds are checked statically
* `[lam|| blah ||]` is user-defined syntax for a literal of type `Expr`; this syntax may be more convenient than writing lots of `Fun`, `Add` and `Var` nodes.

Remember that in every case, we have an alternative that works today, namely 
* `$$(qq "unicode 78")`
* `$$(qq1 "abcdefgh")`
* `$$(checkedNum "341")`
* `$$(lam "blah")`

Is the quasiquote version sufficiently better?  Beyond the name of the quasiquoter, the syntactic baggage is
* 6 characters for an explicit splice, namely two dollars, two parens, two quotes
* 6 characters for a quasiquote, namely open and close brackets and four vertical bars.

Here I'm assuming that the _typed_ version of quasiquotation uses two vertical bars, just like the type version of TH quotes.  I suppose that we could be clever, and switch between typed and untyped quasiquoters based on the type of the quasiquoter, but that's uncomfortable. (The untyped ones run in the renamer which knows nothing of types.)

Either way, the difference is not large.  Question: does this proposal really pay its way?

One might ask the same question of the _existing_ quasiquotes, which can be similarly replaced. with splices.  Back in 2007 when they were designed, all TH splices ran in the typechecker, and quasiquotes (which must run in the renamer) were something new.   But nowadays, with the clear untyped-TH vs typed-TH split, I think we might well reject quasiquotes as a language extension too.   Maybe I should write a proposal to deprecate quasiquotes!

TL;DR, this proposal makes the language (and its implementation) just a bit more complicated, in exchange for what amounts to a minor syntactic variation in concrete syntax.   I have not heard anyone saying "my life woud be so much better with this".   

So if it were me I wouldn't do it.   But if everyone else is keen I won't stand in the way.   The strongest argument is that (for better or worse) we do currently have quasiquotes, and this adds them to the typed-TH box as well as the untyped-TH box.

If it is accepted, could the proposal be modified:
* Stress that this only applies to _expression_ quasquotes.   For patterns, types, and declarations you only have the untyped version.
* Make explicit the concrete syntax: two vertical bars or one?

parsonsmatt commented:
Quasioquotes are nice because they don't deal with weird escaping rules, work great over multiple lines, and are syntactically much easier. Replacing `[quote| ... |]` with `$$(quote "...")` would dramatically harm almost every use of QQ that I have.

simonpj commented:
That's a very helpful observation, thanks Matt.   But it's kind of accidental.    You are really saying that Haskell lacks a good notation for string literals, especially
* Line breaks
* The need to escape double quotes

and it happens that the way that quasiquotes delimits strings is better for very long strings.  Nothing directly to do with quasiquotation at all.

Indeed `[pure|| blah]` :: String, where `pure:: String -> Code String` is perhaps a useful way to get a _string_ literal!   Instead of `"foo \"bar\""` you could write `[pure|| foo "bar"|]`. 

It's really good to hear from people using this stuff in the field.   More please!

bitemyapp commented:
@simonpj https://gitlab.com/lorepub/moot/blob/master/src/Model.hs#L28-135

This is how the database models have been defined on almost every project I've worked on in the last 5 and 1/2 years, save the one that was Opaleye.

I also use quasiquoters for:

- [HTML templates](https://gitlab.com/lorepub/moot/blob/master/src/Handler/Home.hs#L77-94)
- raw string literals with interpolation (think: [format!](https://doc.rust-lang.org/std/macro.format.html))
- [Type-safe HTTP application routes](https://gitlab.com/lorepub/moot/blob/master/src/Routes.hs#L9-49)

I'm not really sure what you're angling at but it unnerves me.

simonpj commented:
> I'm not really sure what you're angling at but it unnerves me.

Don't be un-nerved!  I'm just pointing out that, in principle, a quasiquote `[foo| blah|]`  can always be written `$(foo "blah")`, which is not so very different.

But Matt points out that a quote string `"..."` has different multi-line and escape conventions to the string captured by a quasiquote `[|| ...|]`.   And in practice that makes a very big difference indeed.

I had not fully appreciated that because I'm not an at-scale user of quasiquotation, so I was expressing apprection of some concrete examples from the front line, which you have helpufully added to.

No more than that.   To me, this says that we should have designed a better notation for string literals in the first place!

gridaphobe commented:
Indeed, quasiquoters are so much nicer than our notation for multi-line string literals that I've seen at least one package that provides a quasiquoter that is just for string literals (e.g. the `q` quasiquoter from https://hackage.haskell.org/package/interpolatedstring-perl6-1.0.1/docs/Text-InterpolatedString-Perl6.html).

Ericson2314 commented:
👍 to these recent messages. I maintain this proposal is good *while* we have quasiquoters, but I actually despise them and would love to see them removed:

- As mentioned, they elide the natural separation of concerns between parsing and splicing.

- The interface is a rather ad-hoc record. E.g. what happens if we allow splicing in more positions?

- The name is idiotic! *Quasi*quotation is so called because the quote can be escaped with an unquote. We have that already with regular quotes and splicies. Quasiquoters don't have an unquote/splice construction so they are *less* "quasi" than the baseline. It's madness! The only explanation I've heard is Perl and other languages made the mistake first, so at least we're following some precedent albeit a rotten one.

isovector commented:
@simonpj do you have any further concerns about this proposal? If not, I'll be happy to make the desired changes to the proposal text --- namely the double-bar syntax and that `TQuasiQuoter a`s may have constraints on `a` that participate in usual typechecking machinery both at their definition and splice sites.

simonpj commented:
> and that TQuasiQuoter as may have constraints on a that participate in usual typechecking machinery both at their definition and splice sites.

I couldn't parse that sentence.   Perhaps you can make the changes an point us all at the relevant bit?

But generally, yes ok.

goldfirere commented:
I think this discussion is very interesting. If we didn't have back-compat to worry about, I would probably advocate strongly for a new, better string-literal syntax (allowing multiline literals without escapes) and to remove quasiquotes. (Another infelicity about quasiquotes: they're much more like splices than quotes, yet both the name and syntax make them look like quotes.)

Of course, we *do* have back-compat to worry about, and so I don't think I would do this. And, on balance, I support this proposal as it fills out a missing box. But it's good to know that, almost the only reason to have quasiquotes is because we lack a better string-literal syntax, something I don't think I would have jumped to.

isovector commented:
I've updated the proposal here: https://github.com/ghc-proposals/ghc-proposals/blob/eeeee74fdfd745a7906d237be112715323a7f887/proposals/0000-type-annotated-quoters.md . Can't for the life of me figure out how to merge it into this PR. I intend to merge this into `master` on Wednesday, July 31 if nobody has any complaints with my rewording.

/remind me to wants to accept this on July 31

isovector commented:
@nomeata looks like we've reached consensus here and are ready to accept!

nomeata commented:
> @nomeata looks like we've reached consensus here and are ready to accept!

Done. Would you care to also notify the mailing list?

Ericson2314 commented:
FWIW I wrote https://github.com/ghc-proposals/ghc-proposals/issues/260 because while I approve of this proposal in the short term, I would like to see QuasiQuotes deprecated in the long term.
